server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,734 @@
+        this.replicaShardBatchAllocator = new InternalReplicaBatchShardAllocator();
+        this.batchStartedAction = batchStartedAction;
+        this.batchStoreAction = batchStoreAction;
+        this.maxBatchSize = GATEWAY_ALLOCATOR_BATCH_SIZE.get(settings);
+    }
+
+    @Override
+    public void cleanCaches() {
+        Releasables.close(
+            batchIdToStartedShardBatch.values().stream().map(shardsBatch -> shardsBatch.asyncBatch).collect(Collectors.toList())
+        );
+        batchIdToStartedShardBatch.clear();
+        Releasables.close(
+            batchIdToStoreShardBatch.values().stream().map(shardsBatch -> shardsBatch.asyncBatch).collect(Collectors.toList())
+        );
+        batchIdToStoreShardBatch.clear();
+    }
+
+    // for tests
+    protected ShardsBatchGatewayAllocator() {
+        this.rerouteService = null;
+        this.batchStartedAction = null;
+        this.primaryShardBatchAllocator = null;
+        this.batchStoreAction = null;
+        this.replicaShardBatchAllocator = null;
+        this.maxBatchSize = DEFAULT_SHARD_BATCH_SIZE;
+    }
+
+    // for tests
+
+    @Override
+    public int getNumberOfInFlightFetches() {
+        int count = 0;
+        for (ShardsBatch batch : batchIdToStartedShardBatch.values()) {
+            count += (batch.getNumberOfInFlightFetches() * batch.getBatchedShards().size());
+        }
+        for (ShardsBatch batch : batchIdToStoreShardBatch.values()) {
+            count += (batch.getNumberOfInFlightFetches() * batch.getBatchedShards().size());
+        }
+
+        return count;
+    }
+
+    @Override
+    public void applyStartedShards(final List<ShardRouting> startedShards, final RoutingAllocation allocation) {
+        for (ShardRouting startedShard : startedShards) {
+            safelyRemoveShardFromBothBatch(startedShard);
+        }
+    }
+
+    @Override
+    public void applyFailedShards(final List<FailedShard> failedShards, final RoutingAllocation allocation) {
+        for (FailedShard failedShard : failedShards) {
+            safelyRemoveShardFromBothBatch(failedShard.getRoutingEntry());
+        }
+    }
+
+    @Override
+    public void beforeAllocation(final RoutingAllocation allocation) {
+        assert primaryShardBatchAllocator != null;
+        assert replicaShardBatchAllocator != null;
+        ensureAsyncFetchStorePrimaryRecency(allocation);
+    }
+
+    @Override
+    public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {
+        assert replicaShardBatchAllocator != null;
+        List<List<ShardRouting>> storedShardBatches = batchIdToStoreShardBatch.values()
+            .stream()
+            .map(ShardsBatch::getBatchedShardRoutings)
+            .collect(Collectors.toList());
+        if (allocation.routingNodes().hasInactiveShards()) {
+            // cancel existing recoveries if we have a better match
+            replicaShardBatchAllocator.processExistingRecoveries(allocation, storedShardBatches);
```

Comment:
why do we need to pass the batches explicitly here?

Reply:
fetchData happens inside `processExistingRecoveries`, that require shards from specific batches only as fetchData is called for 1 batch at a time.